Feat/webpack cachebusting

Adezandee opened the pull request:
Cache busting for CSS files + Minify HTML and CSS. #142

Indigo744 commented:
Nice! 😄 

Some questions: how does it works exactly? I see you added the `build:dev` command.
(maybe we should update the [README.md](https://github.com/lyrgard/ffbeEquip/blob/master/README.md) for dev environment setup).

In the webpack config, I see that you identify image with the `file-loader` in /static, but what about image loaded at runtime by js modules?

Adezandee commented:
@Indigo744 
It does not change the dev workflow. I added a `build:dev` command because you asked for it :D but it's not necessary, for now.
The only thing it does is:
- copy html files from `/static/` to `/dist/`, replacing CSS import with hashed urls for cache busting + some html minify
- copy imported CSS files to `/dist/` applying `cssnano` minify

I've changed the server to "shadow" `/static` with `/dist` if files are built

Adezandee commented:
Image files a untouched. Nothing changes HTML or dynamic JS.

Indigo744 commented:
Ah ok, I get it 😸  very nice!

Next step is JS + minified ?

Adezandee commented:
Added JS assets cache busting

lyrgard commented:
What is the app.js that is generated in /dist ?

Adezandee commented:
@lyrgard 
That an unused build artefact. Webpack is designed to compile a full project into one `app.js` file. With the long term objective to use a frontend framework to rewrite the app or part of the app, it made sense to use that tool (and it works great for todays needs too).
I could not find a way to delete the build result since its intended to be the main result.

Adezandee commented:
@lyrgard 
yeah a nice QoL feature for devs :) not required to add `?version=X` anymore

lyrgard commented:
It's working great ! Thanks a LOT for your QoL additions !

Adezandee commented on webpack.config.js:
```diff
@@ -0,0 +1,42 @@
+      {
+        test: /\.(js)$/,
+        use: [
+          { loader: 'file-loader', options: { name: '[path][name].[ext]?[hash:8]', context: './static/', emitFile: false } },
```
yes, `js` files are untouched (no minify step). So I dont emit them.
Same for images files

Indigo744 replied:
So there is no cache busting for JS files, isn't it?
Because the commit description `feat(webpack): js assets cachebusting` suggested otherwise...

Adezandee replied:
There is cache busting for `js` files. You can check on the live website:
view-source:https://ffbeequip.com/#e30=

Indigo744 replied:
Ah I get it. I don't know why I though files needed to be in `dist` to be cached... Nevermind.

Do you plan to add JS minifying? Or can I start looking?

I'm currently looking for cache headers improvements.

Adezandee replied:
Sure.
I think to really improve cache headers you'll have serve assets with `apache`.
Now that everything can be modified easily:
- asset path (ie. base path is now `/`, except in dynamic JS unfortunately).
- apache config (see ansible)

You'll have to catch an special `/assets/` route dans serve files.

Indigo744 replied:
Yup. Using `public, max-age=0` for HTML since they can't be cache-busted :wink: thanks for the reminder.
